python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -34,7 +34,7 @@ def take_table(
     if any(_is_column_extension_type(col) for col in table.columns):
         new_cols = []
         for col in table.columns:
-            if _is_column_extension_type(col):
+            if _is_column_extension_type(col) and col.num_chunks > 1:
```

Comment:
This provides most of the speedup (0.03GBps -> 0.12GBps on the attached benchmark).

Reply:
another question. IIUC, here we will concatenate all the chunks for the entire column, which is not necessary. Because the `indices` may only fall into some of the chunks. We can potentially optimize this by only taking the rows from partial  chunks. Is my understanding correct?

Reply:
I think it's actually always 1 since we only ever take from a fully concatenated table, which coalesces the chunks already (batcher.py:326):

```
            # Build the new shuffle buffer.
            self._shuffle_buffer = self._builder.build()
            if (
                isinstance(
                    BlockAccessor.for_block(self._shuffle_buffer), ArrowBlockAccessor
                )
                and self._shuffle_buffer.num_columns > 0
                and self._shuffle_buffer.column(0).num_chunks
                >= MIN_NUM_CHUNKS_TO_TRIGGER_COMBINE_CHUNKS
            ):
                self._shuffle_buffer = transform_pyarrow.combine_chunks(
                    self._shuffle_buffer
                )
```

This is a bit brittle though, not sure we can avoid regressions without a microbenchmark.

> another question. IIUC, here we will concatenate all the chunks for the entire column, which is not necessary. Because the indices may only fall into some of the chunks. We can potentially optimize this by only taking the rows from partial chunks. Is my understanding correct?

I believe so. Do you think that approach would be preferable?

Reply:
I'm still very confused about the logic of `ShufflingBatcher.next_batch`. Especially the following part.
Why put the shuffle_buffer back to the builder, while shuffle_buffer is built from the builder? This seems to copy data back and forth again and again. This could hurt perf badly when we have large `shuffle_buffer_min_size` and small `batch_size`.

```python
if self._batch_head > 0:
    # Compact the materialized shuffle buffer.
    # TODO(Clark): If alternating between adding blocks and fetching
    # shuffled batches, this aggressive compaction could be inefficient.
    self._shuffle_buffer = BlockAccessor.for_block(
        self._shuffle_buffer
    ).take(self._shuffle_indices[self._batch_head :])
# Add the unyielded rows from the existing shuffle buffer.
self._builder.add_block(self._shuffle_buffer)
```

Reply:
The process in my head is very simple:
1. `add` keeps adding data to the builder. 
2. when builder size reaches `shuffle_buffer_min_size`, it builds the shuffle buffer and shuffles the order (this will copy shuffle_buffer_min_size data).
3. `next_batch` just takes a batch from the shuffle_buffer (this only copies batch_size data).

while the current logic seems too complex. there are a lot of copies (take, add_block, combine_chunks).

Reply:
Yeah I think I agree. I'll try to rewrite this class to simplify

Reply:
I ended up just removing `shuffle_indices` from the current implementation, keeping the compaction logic as we still need that for performance in any case. Switching from take() with shuffle indices to slice() with up-front random shuffle further improves performance from `0.2 GBps` to `0.3 GBps`.